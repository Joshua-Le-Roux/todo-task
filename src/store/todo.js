import { createSlice } from "@reduxjs/toolkit";


export const todoSlice = createSlice ({
    name: "todo",
    initialState: {
        nextId: 2,
        data:
        {
            1:{
                content: "Clean kitchen",
                completed: false
            },

        }
        
    },

    reducers: {
        
        edit: (state, action) => {
            // creating action payload id to grab specific task ids
            let todoId = action.payload.id
            if(state.data[todoId]) {
                // accessing particular state element
                state.data[todoId] = {
                    //assigning the user input taken from prompt as the new state
                    content : action.payload.content,
                    completed: false
                }
            } else {
            }
        },

        completed: (state, action) => {
            let todoId = action.payload.id
            //taking the id to mark the specific action as either true or false
            state.data[todoId].completed = action.payload.completed
        },

        remove: (state,action) => {
            let todoId = action.payload
            //using the id and deleting the specific data element
            delete state.data[todoId]
        },

        newTodo: (state, action) => {
            //assigning the next id from initialState
            let newId = state.nextId
            //creating new data element 
            state.data[newId] = {
                // taking the input from the input field and setting that as the content
                content : action.payload,
                completed : false,
            }
            //adding 1 on to the nextId for the next added task
            state.nextId++;
        },
    },
});

export const { edit, completed, remove, newTodo } = todoSlice.actions;

 export default todoSlice.reducer;