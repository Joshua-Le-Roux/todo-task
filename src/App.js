
import './App.css';
import { useDispatch, useSelector } from 'react-redux/es/exports';
import { newTodo } from './store/todo';
import { useState } from 'react';
import Buttons from "./buttons"

function App() {
  const [weather, setWeather] = useState(null)
  //useState to access the user input and assign it to the state
  const [addTodo, setaddTodo] = useState('')
  //calling dispatch
  const dispatch = useDispatch();

  fetch("https://api.open-meteo.com/v1/forecast?latitude=-33.9258&longitude=18.4232&hourly=temperature_2m,relativehumidity_2m,windspeed_10m&current_weather=true")
  //converting from text to obj
  .then(res => res.json())
  .then((result)=> {
      //grabbing the current temp in
      let currentTemp = result.current_weather.temperature
          // returning current temp in Paris
      setWeather(currentTemp)
  }),
  (error) => {
      console.log(error)
  }

  const submitAdd = (event) => {
    event.preventDefault();
    //dispatching addTodo with newTodo reducer
    dispatch(newTodo(addTodo))
    setaddTodo('')
  };

  //grabbing the todo data to make array to iterate over each item in todo.data
  const todoItems = useSelector((state) => state.todo.data)
  

  return (
    <div className="App">
      {weather && <h2>Cape Town : {weather} C</h2>}
      <form>
        <label>Add Task<br/>  
        {/* when input field is changed its change is monitored and then set to the useState */}
          <input type='text' value={addTodo} onChange={(e) => setaddTodo(e.target.value)} />
        </label><br/>
        {/* when add is clicked the user input is taken and a new todo is created */}
        <button type='submit' onClick={submitAdd}>add</button>
      </form><hr/>
      <div>
        {/* using Object.entries on previously created todoItems to create an array to iterate and display all tasks with their own buttons */}
        {Object.entries(todoItems).map(([todoId, todo]) => (
            <div key={todoId}>
              Task : {todo.content} <br/> Completed : {todo.completed ? "True" : "False"}
              <Buttons id={todoId} complete={todo.completed} />
              <hr/>
            </div>
        ))}
      </div>
    </div>
  );
};

export default App;
