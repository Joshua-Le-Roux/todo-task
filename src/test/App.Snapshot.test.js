import React from 'react';
import App from '../App'
import renderer from "react-test-renderer"
import { Provider } from 'react-redux';
import store from "../store/store"
// import { useSelector } from 'react-redux/es/hooks/useSelector';

    test('matches snapshot', () => {
    const tree = renderer
    .create(<Provider store={store}><App /></Provider>)
    .toJSON();
    expect(tree).toMatchSnapshot();
})