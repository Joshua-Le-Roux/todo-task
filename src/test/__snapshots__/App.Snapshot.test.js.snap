// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  className="App"
>
  <div
    onLoad={[Function]}
  />
  <form>
    <label>
      Add Task
      <br />
      <input
        onChange={[Function]}
        type="text"
        value=""
      />
    </label>
    <br />
    <button
      onClick={[Function]}
      type="submit"
    >
      add
    </button>
  </form>
  <hr />
  <div>
    <div>
      Task : 
      Clean kitchen
       
      <br />
       Completed : 
      False
      <div>
        <button
          onClick={[Function]}
        >
          Edit
        </button>
        <button
          onClick={[Function]}
        >
          Completed
        </button>
        <button
          onClick={[Function]}
        >
          Delete
        </button>
      </div>
      <hr />
    </div>
  </div>
</div>
`;
